### ISSUE TITLE
Docker daemon not recognising containers on other hosts

### ISSUE THREAD

**Issue by YP28** [coreos/flannel#281]
Correct me if I'm wrong, but the way I understood flannel was for, was to enable cross host container linking like this:

``` bash
docker run --name service \
    --link src-container:dest-mapping \
    ubuntu /bin/bash
```

This would require flannel to be setup to create subnets on each host, which will then enable container linking through the container assigned subnet IP addresses by running docker with the `--bip` variable with the subnet generated by flannel.

However, I cannot get this to work. Again, I'm not sure if I understand the purpose of flannel completely, but if my understanding is correct, I must be missing something.

I've created a section in my cloud-config to enable and start the flannel daemon:

``` yaml
  - name: flanneld.service
    drop-ins:
      - name: 50-network-config.conf
        content: |
          [Service]
          ExecStartPre=/usr/bin/etcdctl set /coreos.com/network/config '{ "Network": "172.0.0.0/8", "SubnetLen": 20, "SubnetMin": "172.10.0.0", "SubnetMax": "172.99.0.0", "Backend": { "Type": "udp", "Port": 7890 } }'
    command: start
```

I've checked whether the configuration gets written to etcd, and to my knowledge this means the subnets are set up for the hosts:

``` bash
$ etcdctl get /coreos.com/network/config
# { "Network": "172.0.0.0/8", "SubnetLen": 20, "SubnetMin": "172.10.0.0", "SubnetMax": "172.99.0.0", "Backend": { "Type": "udp", "Port": 7890 } }
$ etcdctl ls --recursive /coreos.com/network
# /coreos.com/network/config
# /coreos.com/network/subnets
# /coreos.com/network/subnets/172.10.144.0-20
# /coreos.com/network/subnets/172.15.96.0-20
# /coreos.com/network/subnets/172.13.16.0-20
```

I've also setup docker to use the subnet when running the docker daemon:

``` yaml
  - name: docker.service
    drop-ins:
      - name: 50-docker-bip.conf
        content: |
          [Service]
          EnvironmentFile=/run/flannel/subnet.env
          Environment=DOCKER_OPT_BIP='--bip=${FLANNEL_SUBNET}'
          Environment=DOCKER_OPT_MTU='--mtu=${FLANNEL_MTU}'
          Environment=DOCKER_OPT_IPMASQ='--ip-masq=${FLANNEL_IPMASQ}'
```

When I check if docker picks up the configuration and uses it by executing `systemctl status docker` the output is:

``` bash
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib64/systemd/system/docker.service; disabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─50-docker-bip.conf
   Active: active (running) since Mon 2015-08-10 21:51:38 UTC; 19h ago
     Docs: http://docs.docker.com
 Main PID: 964 (docker)
   Memory: 351.9M
      CPU: 45.075s
   CGroup: /system.slice/docker.service
           └─964 docker --daemon --host=fd:// --bip=172.15.96.1/20 --mtu=1472 --ip-masq=false
```

And when I check if the daemon actually runs with that configuration by executing `ps aux | grep "docker --daemon` one of the entries confirms that it does:

``` bash
root       964  0.0  2.2 585512 23028 ?        Ssl  Aug10   0:41 docker --daemon --host=fd:// --bip=172.15.96.1/20 --mtu=1472 --ip-masq=false
```

Starting containers on separate hosts and checking the interface configuration confirms that the docker containers do, as expected, get an IP address from the subnet pool assigned to them and pinging indicates that they can reach each other over the new container network:

``` bash
$ root@897c201cf2e7:/# ifconfig
# inet addr:172.10.144.4  Bcast:0.0.0.0  Mask:255.255.240.0

$ root@897c201cf2e7:/# ping 172.15.96.4
# PING 172.15.96.4 (172.15.96.4) 56(84) bytes of data.
#64 bytes from 172.15.96.4: icmp_seq=1 ttl=60 time=2.82 ms
```

``` bash
$ root@ab574c69915c:/# ifconfig
# inet addr:172.15.96.4  Bcast:0.0.0.0  Mask:255.255.240.0

$ root@ab574c69915c:/# ping 172.10.144.4
# PING 172.10.144.4 (172.10.144.4) 56(84) bytes of data.
#64 bytes from 172.10.144.4: icmp_seq=1 ttl=60 time=2.65 ms
```

Okay, all looks fine. So now my problem, I'm trying to link a new container to an existing container running on an other host. To verify the container is running, I run `fleetctl list-units`:

``` bash
UNIT                        MACHINE                     ACTIVE  SUB
authentication-db.service   4a6f990f.../188.226.248.43  active  running
```

To keep things easy, the docker container the service is running is also called 'authentication-db' using `--name authentication-db` in the docker run command.

Now I try to run `docker run -i -t --link authentication-db:database ubuntu /bin/bash`, but it gives me the following error:

``` bash
FATA[0000] Error response from daemon: Could not get container for authentication-db
```

What I read from other posts one of the problems might be that the flannel0 or docker0 interface does not use the correct default gateway, thus preventing it to connect to the right containers. I'm not sure, but when I run `route -n` the only default gateway is a public IP address. As I'm not very familiar with this subject, I'm not sure whether this configuration is correct or not:

``` bash
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         95.85.26.1      0.0.0.0         UG    0      0        0 eth0
0.0.0.0         0.0.0.0         0.0.0.0         U     2048   0        0 eth0
10.129.0.0      0.0.0.0         255.255.0.0     U     0      0        0 eth1
95.85.26.0      0.0.0.0         255.255.255.0   U     0      0        0 eth0
169.254.0.0     0.0.0.0         255.255.0.0     U     0      0        0 eth0
172.0.0.0       0.0.0.0         255.0.0.0       U     0      0        0 flannel0
172.15.96.0     0.0.0.0         255.255.240.0   U     0      0        0 docker0
```

I'm kinda lost as I don't know where to find the solution to this problem, and if this is even the purpose of the flannel service.

**Comment by eyakubovich** [comment]
@YP28 You are correct that flannel was designed for container-to-container communication across hosts by giving each container an unique and routable IP within your cluster. However it was not designed to make links work across nodes. Docker links really provide service discovery via environment variables and /etc/hosts. Since flannel does not make links visible across hosts, you need to use some other way to do service discovery. There're a couple of ways to do that. You can just put values into etcd under some key that signifies your service. Or you can use [SkyDNS](https://github.com/skynetservices/skydns) which is a DNS server backed by etcd.

You were able to ping one container from another on different hosts by IP without mapping ports -- that's exactly what flannel makes possible.

Since you're using flannel on CoreOS, it is a lot simpler to set up than what you went through. Please see this guide https://coreos.com/flannel/docs/latest/flannel-config.html

**Comment by YP28** [comment]
Thanks for your feedback @eyakubovich. I used that guide at first but modified it to fit my case (DigitalOcean uses the 10.0.0.0 address space) and tried some things out to test it.

So I did set it up properly. What I'm wondering though, is it comparable to this: https://docs.docker.com/articles/networking/#docker-ipv6-cluster
Or does that do enable cross-host linking of containers? I can't really get that from the documentation and I'm guessing you looked into that before creating flannel? 

If that is not the same based on ipv6, and does implement the ability to use the native docker link parameter to link containers, why not implement the ipv6 networking in flannel to enable it? And if it is the same as what flannel does, why didn't Docker choose to just use ipv4?
